app/models/account_migration.rb

```diff
@@ -0,0 +1,155 @@
+class AccountMigration < ApplicationRecord
+  include Diaspora::Federated::Base
+
+  belongs_to :old_person, class_name: "Person"
+  belongs_to :new_person, class_name: "Person"
+
+  validates :old_person, uniqueness: true
+  validates :new_person, uniqueness: true
+
+  # TODO: needs spec
+  after_create :lock_old_user!
+
+  attr_accessor :old_private_key
+
+  def receive(*)
+    perform!
+  end
+
+  def public?
+    true
+  end
+
+  def sender
+    @sender ||= old_user || ephemeral_sender
+  end
+
+  # Normally pod initiates migration locally when the new user is local. Then the pod creates AccountMigration object
+  # itself. If new user is remote, then AccountMigration object is normally received via the federation and this is
+  # remote initiation then.
+  def remotely_initiated?
+    new_person.remote?
+  end
+
+  def locally_initiated?
+    !remotely_initiated?
+  end
+
+  # executes a migration plan according to this AccountMigration object
+  def perform!
+    ActiveRecord::Base.transaction do
+      account_deleter.tombstone_person_and_profile
+      account_deleter.close_user if user_left_our_pod?
+      account_deleter.tombstone_user if user_changed_id_locally?
+
+      update_all_references
+    end
+
+    dispatch if locally_initiated?
+    dispatch_contacts if remotely_initiated?
+  end
+
+  # TODO: docs
+  def subscribers
+    subscribers = new_user.profile.subscribers.remote
+    return subscribers if old_person.local?
```

Comment:
Nit: I think with a tap (similar like [here](https://github.com/diaspora/diaspora/blob/d0b634852b440a8c3deac9734270d8577ef5a06b/app/models/post.rb#L151-L153)) this is more readable.